### ISSUE TITLE
[Feature request] App UI language option.

### ISSUE THREAD

**Issue by c933103** [MarcusWolschon/osmeditor4android#915]
According to my understanding, currently, the Vespucci app would automatically detect the user language setting in the device that the app have been installed on, and will display the same language as that setting if the language is available, which mean as I am a native Chinese user and I have configured my phone to use traditional Chinese, the app would automatically display Traditional Chinese text as the default user interface language to me after I installed the app.

However, when I tried to use the user app, I found that the way they are translated are different from the way I would normally use/read/write the Chinese language and thus they are really hard for me to understand what each individual options in the app mean. I tried to switch the app language to English so that it would be easier for me to use the app, however I cannot find such option in the app user interface menu. Could you guys add such an option to the app setting so that I can use English UI instead of the Chinese UI, without turning my entire phone into English?

**Comment by simonpoole** [comment]
Hi, thanks for the report.

Android doesn't provide an easy (that is non-hackish) way to have per app language selection (we do provide a way to use English for the presets, but the preset translation system is independent of Android). 

That said I suspect the issue is simply to a mix up of language codes, because obviously if you are getting wrong/weird translations we should simply fix that instead of adding yet another option. Do you know what language code your device is using? And can you identify the weird translations here https://www.transifex.com/openstreetmap/vespucci/dashboard/ ?

**Comment by c933103** [comment]
My phone setting language is zh-HK while the language used by the app should be from zh-TW translation. Those weirdness probably come from the regional difference in Chinese language. So it seems like to solve them directly, another translation branch would have to been made to cater the regional difference. But then it seems like it mean a thousand of string would need to be translated again and that seems to be quite a lot of work and future updates will also be needed. To me it seems like just let me use English would be an easier solution.

p.s. how about https://stackoverflow.com/questions/12908289/how-to-change-language-of-app-when-user-selects-language ?

**Comment by simonpoole** [comment]
> 
> 
> My phone setting language is zh-HK while the language used by the app should be from zh-TW translation. Those weirdness probably come from the regional difference in Chinese language. So it seems like to solve them directly, another translation branch would have to been made to cater the regional difference. But then it seems like it mean a thousand of string would need to be translated again and that seems to be quite a lot of work and future updates will also be needed. To me it seems like just let me use English would be an easier solution.

The only question is if there are any significant differences between zh-HK and zh-TW. If no we can simply duplicate the file, if yes we can copy the existing translations to zh-HK and allow for slightly differing translations.

> 
> p.s. how about https://stackoverflow.com/questions/12908289/how-to-change-language-of-app-when-user-selects-language ?

The answers are from 2012 and as such very outdated (complain to google that this is so difficult, not to us, thank you).

**Comment by simonpoole** [comment]
I've duplicated the translated files in transifex and added a language mapping in https://github.com/MarcusWolschon/osmeditor4android/commit/35fa20ee0a2b67ed36fb59d8c75158d9aa46386c

**Comment by danieldjewell** [comment]
FWIW, I would +1 this ... There are many cases where the Android locale might not match the locale in which you want to use the app... (it also means that maintaining the translations is that much more important -- as @c933103 points out, English is a good fallback.) 

That said, https://stackoverflow.com/questions/2900023/change-app-language-programmatically-in-android has some much more recent answers and suggestions - I'm unsure which one would be best, but it appears that there are some relatively straightforward options. 

I'm not a Chinese speaker but as @c993103 indicates, there might be significant regional differences. There are many cases, worldwide, where someone might have the OS Locale set one way but feel more comfortable in another language. Given the breadth and complexity of the feature set in Vespucci (there are pretty technical terms in the settings!), it seems extra-important that someone would be able to select the language they feel most comfortable using regardless of the OS Locale. 

(@simonpoole you're in ZH-CH right? Just think of the differences between de-CH and de-DE -- granted, in writing, it's not a problem -- but say someone wanted to do a Swiss German written version - like DuckDuckGo did. Without the ability to override, could be problematic. Also, even just in CH -- maybe someone has their phone set in fr-CH because they are learning French but maybe, because of the technical nature of Vespucci, they'd feel more comfortable using it in de-CH or de-DE ....        ) 

Merci vilmal! ;)

**Comment by cvigo** [comment]
The issue for me is that Spanish translation is terrible (guess Google translator) and many many words are absolutely misleading if not hilarious. Many apps which translation is in progress and based on community contributions have the choice of selecting the UI language independently of the language user setting in Android. I think that is what's being asked for here.

**Comment by simonpoole** [comment]
@cvigo why don't you fix the translation then? 

That has the benefit of improving the situation for everybody that would like to use the ES translation and not EN. There are only a total of 975 strings in the UI so it is not a big thing (and lots of the existing translations would seem to be OK).

**Comment by cvigo** [comment]
@simonpoole agree, I did several fixes today. In fact most of them in OSM resources. Hope to see them soon

**Comment by georg-d** [comment]
Having  the possibility to switch to EN (not other languages) would allow users to see/note/record button & menu & option names exactly as the Vespucci UI has in EN. This makes bug reports, enhancement requests etc. much easier to understand & reproduce compared to ad-hoc translations, texts in EN but screenshots in ZH/ES/DE/...
 
Besides that, personally, I prefer to have one consistent language insteadt of a mix of EN (e.g. most OSM keys & values) and DE (many UI elements).

**Comment by leojth** [comment]
+1
I would also want to have an option to switch to english. I know all the technical terms in english and prefer to work with OSM in english. I also have my primary editor, JOSM, set to english. You're doing a really great job with Vespucci, but my language, swedish, is poorly translated and I think the translation is doing more harm then help at the moment (its really confusing for me). Swedish is as small language and I think  it's hard for small languages to keep up with changes made in specialized apps like Vespucci. Better to lay focus on the big languages. So keep up the good work! ... and please, give me an english option :-)

**Comment by zedlabs** [comment]
Hi Simon, as I was trying to figure out a solution to this problem I found[ this library](https://github.com/zeugma-solutions/locale-helper-android), I implemented it in the app and it seems to work well. 
Only change required will be to make all the activities extend `LocaleAwareCompatActivity` and the application class to `LocaleAwareApplication`
I'm new to the codebase so I don't if this would be sufficient to cover the entire application, please let me know what you think about this

**Comment by simonpoole** [comment]
@zedlabs a well packaged hack still remains a hack naturally :-) and doesn't stop it from being something google doesn't want done. There are a few libs that provide for this, though in the end outside of providing some mechanism to store the language as a preference, it's is rather trivial. 

What would need to be checked for the specific lib is if there is any noticeable impact from the dependency on the Kotlin standard library (probably not, but rather safe than sorry) and any API 29/30 differences that might cause issues.

**Comment by sorz** [comment]
Quick fix w/o any hacking: just remove low-quality translations.

Set up some quality requirement on locales (e.g.  reviewed strings > x% and/or untranslated strings < y%), only import these languages which meet the requirement from Transifex.

Or, make the inclusion a manual process: create issues before include any locale, let we the community discuss whether it is good enough to use.

**Comment by c933103** [comment]
> Quick fix w/o any hacking: just remove low-quality translations.
> 
> Set up some quality requirement on locales (e.g.  reviewed strings > x% and/or untranslated strings < y%), only import these languages which meet the requirement from Transifex.
> 
> Or, make the inclusion a manual process: create issues before include any locale, let we the community discuss whether it is good enough to use.
> 

It won't solve the issue, since what is "hard to understand" for a language's user in Region A, could be common day-to-day expression for user of same language in Region B

**Comment by sorz** [comment]
I mean let them fallback to English.
Yes it doesn't resolve all the issue, such as zh-HK people want zh-TW, and de-CH -> de-DE as per discussion.
But the quality of some locales are so low that I think we can let them fallback to English safely, which is better than what it is now.
Perhaps I should create a separated issue for this. It's somewhat orthogonal to this issue.

**Comment by simonpoole** [comment]
..
> Or, make the inclusion a manual process: create issues before include any locale, let we the community discuss whether it is good enough to use.
..
Anybody qualified to discuss the merits of a specific translation could simply fix it. 

The more general question is when is a translation no longer fit for purpose, that is to make it easier for a non-English speaker to use the app. I don't think there is an easy answer to that.

**Comment by LilimVipera** [comment]
Hi!
I'm new to Vespucci and the first thing I searched for were language settings ^^'
I use my OS GB-EN for training purposes, but was quite surprised that such a setting was classified as "hack" - although most of my apps have that option
Especially in slightly tech-y apps like Vespucci I appreciate being able to switch because that way I can compare different localisations and use it to improve understanding 

So, long story short: I'd be very grateful if such an option was available! :)

**Comment by simonpoole** [comment]
That google doesn't want this to be done is completely clear and all the solutions are just workarounds this, so clearly hacks (including that they can break at the drop of a hat, aka googles next version).

**Comment by c933103** [comment]
> That google doesn't want this to be done is completely clear and all the solutions are just workarounds this, so clearly hacks (including that they can break at the drop of a hat, aka googles next version).

Google, being an America-based company, isn't the best in providing localization support to their users around the world.

**Comment by LilimVipera** [comment]
Wow, that's downright frustrating...
Just a second a go the OS itself showed a really good reason for keeping it english, when "appear on top" was translated as "Am Anfang anzeigen"... wow, I didn't remember it to be this crappy

Still, I'd be thankful if a way was found. I found "OSM Go!" on fdroid, they have the option for the most part, and as they're open source maybe it's worth a look?
On the other hand, should I ever stumble upon something seemingly helpful, I'd post it here ;)

**Comment by simonpoole** [comment]
We already have a (currently) working solution which will likely be available in version 16 or so, but you just have to realize that it might break at any point in the future.

**Comment by LilimVipera** [comment]
Sweet!
And breaking apps after updates is something you can even encounter with great AAAs - like Samsungs own music player app that stopped working with Android... 10?
But FOSS-app-creators at least tell you why reasonably (in most cases at least) and normally seem really invested in making something great

So, I don't know about others, but I am grateful that you tried (and seemingly succeeded!) whilst knowing that ol' G can be an ass to break it anyway with every update

In short: Thank you, you're doing a great job! (as far as I can judge as the noob I am ^^')

**Comment by cvigo** [comment]
> That google doesn't want this to be done is completely clear and all the solutions are just workarounds this fact, so clearly hacks (including that they can break at the drop of a hat, aka googles next version).

This is really surprising as Google Maps has an option to select the GUI language 🤷🏻‍♂️

**Comment by simonpoole** [comment]
> 
> This is really surprising as Google Maps has an option to select the GUI language 🤷🏻‍♂️

quod licet Iovi non licet bovi
